include/flashinfer/sampling.cuh

```diff
@@ -2213,6 +2224,712 @@ cudaError_t ChainSpeculativeSampling(DType* draft_probs, IdType* draft_token_ids
+  });
+
+  FLASHINFER_CUDA_CALL(cudaMemcpyAsync(taskLenHost.data(), taskLenPtr[0], sizeof(int) * batch_size,
+                                       cudaMemcpyDefault, stream));
+
+  // === Iter 2 ===
+  DISPATCH_COMPUTE_CAP_NUM_THREADS(compute_capacity, BLOCK_THREADS, {
+    FLASHINFER_CUDA_CALL(cudaStreamSynchronize(stream));
+    int maxTaskLen = *std::max_element(taskLenHost.begin(), taskLenHost.end());
+    int flag = 0;
+
+    if (maxTaskLen != 1) {
+      // clear hist and globalCount
+      FLASHINFER_CUDA_CALL(
+          cudaMemsetAsync(histPtr, 0, sizeof(int) * ((1 << 12) + 1) * batch_size, stream));
+
+      // iter2: countBin
+      dim3 countbin_iter2_nblks((maxTaskLen + BLOCK_THREADS - 1) / BLOCK_THREADS, batch_size);
+      dim3 countbin_iter2_nthrs(BLOCK_THREADS);
+
+      auto countbin_iter2_kernel = countBinKernel<BLOCK_THREADS, 12, 20>;
+      void* countbin_iter2_args[] = {&valBuffer[flag], &taskLenPtr[flag], &histPtr, &d,
+                                     &batch_size};
+      FLASHINFER_CUDA_CALL(cudaLaunchKernel((void*)countbin_iter2_kernel, countbin_iter2_nblks,
+                                            countbin_iter2_nthrs, countbin_iter2_args, 0, stream));
+
+      // iter2: selectBin
+      dim3 selectbin_iter2_nblks(batch_size);
+      dim3 selectbin_iter2_nthrs(BLOCK_THREADS);
+      size_t smem_size = sizeof(SelectBinTempStorage<BLOCK_THREADS>);
+
+      auto selectbin_iter2_kernel = selectBinKernel<BLOCK_THREADS, (1 << 12)>;
+      void* selectbin_iter2_args[] = {&histPtr, &binIdPtr, &kPtr, &taskLenPtr[flag ^ 1]};
+
+      FLASHINFER_CUDA_CALL(cudaFuncSetAttribute(
+          selectbin_iter2_kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, smem_size));
+      FLASHINFER_CUDA_CALL(cudaLaunchKernel((void*)selectbin_iter2_kernel, selectbin_iter2_nblks,
+                                            selectbin_iter2_nthrs, selectbin_iter2_args, smem_size,
+                                            stream));
+
+      // iter2: selectCandidate
+      dim3 selectcan_iter2_nblks((maxTaskLen + BLOCK_THREADS - 1) / BLOCK_THREADS, batch_size);
+      dim3 selectcan_iter2_nthrs(BLOCK_THREADS);
+
+      auto selectcan_iter2_kernel = selectCandidateKernel<BLOCK_THREADS, 12, 20>;
+      void* selectcan_iter2_args[] = {&valBuffer[flag], &valBuffer[flag ^ 1], &globalCountPtr,
+                                      &binIdPtr,        &taskLenPtr[flag],    &d};
+
+      FLASHINFER_CUDA_CALL(cudaLaunchKernel((void*)selectcan_iter2_kernel, selectcan_iter2_nblks,
+                                            selectcan_iter2_nthrs, selectcan_iter2_args, 0,
+                                            stream));
+
+      // update taskLen
+      FLASHINFER_CUDA_CALL(cudaMemcpyAsync(taskLenHost.data(), taskLenPtr[flag ^ 1],
+                                           sizeof(int) * batch_size, cudaMemcpyDefault, stream));
+      flag ^= 1;
+    }
+
+    // === Iter 3 ===
+    FLASHINFER_CUDA_CALL(cudaStreamSynchronize(stream));
```

Comment:
Is it possible to get rid of cudaStreamSynchronization like this?
Seems we rely on host-side maxTaskLen as kernel configuration of `countbin_iter3_kernel`, but is it possible to rewrite `countbin_iter3_kernel` as a persistent kernel (fix grid configuration), but inside the kernel we depend on device-side `maxTaskLen` value (so that we can reduce synchronization overhead here).